### PR TITLE
fix hard reset: reassign r4300_pc_struct after poweron call

### LINKED ISSUE #639: Crash on hard reset using Pure Interpreter/Dynarec
Note: @Gillou68310 confirmed that this happens in standalone as well, so I will create a Issue here for tracking purposes.
Cached interpreter does not crash, but seems to not work either.

Informations as provided by @hhromic :

Crashes with CPU types

    Dynamic Recompiler
    Pure Interpreter

Doesn't crash with CPU types

    Cached Interpreter

Crash using **Dynamic Recompiler** (instrumentalised)

    [INFO] Reset.
    [libretro INFO] mupen64plus-next: Entering retro_reset()
    [libretro INFO] mupen64plus-next: *** M64CMD_RESET ***
    *** Entering main_reset(1) ***
    *** Entering hard_reset_device() ***
    *** Calling call_interrupt_handler(&r4300->cp0, 11) ***
    *** Entering call_interrupt_handler(cp0, 11) ***
    *** Entering reset_hard_handler() ***
    *** Calling poweron_device() ***
    [libretro INFO] mupen64plus-next: Initializing 4 RDRAM modules for a total of 8 MB
    *** Calling pif_bootrom_hle_execute() ***
    *** Calling init_interrupt() ***
    *** Calling invalidate_r4300_cached_code() ***
    *** Calling generic_jump_to(a4000040) ***

    Thread 1 "retroarch" received signal SIGSEGV, Segmentation fault.
    0x64999100 in cached_interp_MTC0 () at mupen64plus-core/src/device/r4300/mips_instructions.def:1185
    1185        switch(rfs)

GDB stack trace using **Dynamic Recompiler**

    Thread 1 (Thread 0x71b41000 (LWP 11149)):
    #0  0x64999100 in cached_interp_MTC0 () at mupen64plus-core/src/device/r4300/mips_instructions.def:1185
            r4300 = 0x6e0f7000 <g_dev>
            cp0_regs = 0x705f8190 <g_dev+38801808>
            cp0_next_interrupt = 0x705f8040 <g_dev+38801472>
            __PRETTY_FUNCTION__ = "cached_interp_MTC0"
    #1  0x64a51430 in MTC0_new (copr=13, count=20480, diff=0, pcaddr=-1543503808) at mupen64plus-core/src/device/r4300/new_dynarec/new_dynarec.c:11208
            r4300 = 0x6e0f7000 <g_dev>
    #2  0x6e5f8024 in g_dev () from /opt/retropie/libretrocores/lr-mupen64plus-next/mupen64plus_next_libretro.so

Crash using **Pure Interpreter** (instrumentalised)

    [INFO] Reset.
    [libretro INFO] mupen64plus-next: Entering retro_reset()
    [libretro INFO] mupen64plus-next: *** M64CMD_RESET ***
    *** Entering main_reset(1) ***
    *** Entering hard_reset_device() ***
    *** Calling call_interrupt_handler(&r4300->cp0, 11) ***
    *** Entering call_interrupt_handler(cp0, 11) ***
    *** Entering reset_hard_handler() ***
    *** Calling poweron_device() ***
    [libretro INFO] mupen64plus-next: Initializing 4 RDRAM modules for a total of 8 MB
    *** Calling pif_bootrom_hle_execute() ***
    *** Calling init_interrupt() ***
    *** Calling invalidate_r4300_cached_code() ***
    *** Calling generic_jump_to(a4000040) ***

    Thread 1 "retroarch" received signal SIGSEGV, Segmentation fault.
    0x649bbd84 in generic_jump_to (r4300=0x6e0f7000 <g_dev>, address=2751463488) at mupen64plus-core/src/device/r4300/r4300_core.c:413
    413             *r4300_pc(r4300) = address;

GDB stack trace using **Pure Interpreter**

    #0  0x649bbd84 in generic_jump_to (r4300=0x6e0f7000 <g_dev>, address=2751463488) at mupen64plus-core/src/device/r4300/r4300_core.c:413
    No locals.
    #1  0x649a75d0 in reset_hard_handler (opaque=0x6e0f7000 <g_dev>) at mupen64plus-core/src/device/r4300/interrupt.c:497
            dev = 0x6e0f7000 <g_dev>
            r4300 = 0x6e0f7000 <g_dev>
    #2  0x649a7688 in call_interrupt_handler (cp0=0x709f86d4 <g_dev+42997460>, index=11) at mupen64plus-core/src/device/r4300/interrupt.c:511
            __PRETTY_FUNCTION__ = "call_interrupt_handler"
            handler = 0x709f8838 <g_dev+42997816>
    #3  0x649a7770 in gen_interrupt (r4300=0x6e0f7000 <g_dev>) at mupen64plus-core/src/device/r4300/interrupt.c:540
            cp0_regs = 0x705f8190 <g_dev+38801808>
            cp0_next_interrupt = 0x705f8040 <g_dev+38801472>
    #4  0x649ae73c in BEQ (r4300=0x6e0f7000 <g_dev>, op=325058571) at mupen64plus-core/src/device/r4300/mips_instructions.def:897
            take_jump = 0
            jump_target = 2148350228
            link_register = 0x705f8080 <g_dev+38801536>
    #5  0x649b9d28 in InterpretOpcode (r4300=0x6e0f7000 <g_dev>) at mupen64plus-core/src/device/r4300/pure_interp.c:393
            op_address = 0x4215a8ec
            op = 325058571
    #6  0x649bb1b4 in run_pure_interpreter (r4300=0x6e0f7000 <g_dev>) at mupen64plus-core/src/device/r4300/pure_interp.c:730
    No locals.
    #7  0x649bb508 in run_r4300 (r4300=0x6e0f7000 <g_dev>) at mupen64plus-core/src/device/r4300/r4300_core.c:137
    No locals.
    #8  0x64982138 in run_device (dev=0x6e0f7000 <g_dev>) at mupen64plus-core/src/device/device.c:252
    No locals.
    #9  0x649c4794 in main_run () at mupen64plus-core/src/main/main.c:1259
            i = 5
            k = 2
            rdram_size = 8388608
            count_per_op = 2
            disable_extra_mem = 0
            si_dma_duration = 2304
            no_compiled_jump = 0
            randomize_interrupt = 0
            eep = {data = 0x712bb3b0 <saved_memory> '\377' <repeats 200 times>..., size = 2048, filename = 0x0}
            fla = {data = 0x712e3bb0 <saved_memory+165888> '\377' <repeats 200 times>..., size = 131072, filename = 0x0}
            sra = {data = 0x712dbbb0 <saved_memory+133120> '\377' <repeats 200 times>..., size = 32768, filename = 0x0}
            dd_rom_size = 0
            dd_disk = {data = 0x0, size = 0, filename = 0x0}
            audio_out_backend_libretro = {set_format = 0x64a59470 <set_audio_format_via_libretro>, push_samples = 0x64a59918 <push_audio_samples_via_libretro>}
            control_ids = {0, 1, 2, 3}
            cin_compats = {{control_id = 0, cont = 0x712b9c20 <g_dev+52177952>, tpk = 0x712b9ce0 <g_dev+52178144>, last_input = 0, last_pak_type = 2, main_switch_pak = 0x0,
                pak_switch_delay = 0, gb_switch_delay = 0, gb_cart_switch_enabled = 0}, {control_id = 1, cont = 0x712b9c38 <g_dev+52177976>, tpk = 0x712b9cf4 <g_dev+52178164>,
                last_input = 0, last_pak_type = 1, main_switch_pak = 0x0, pak_switch_delay = 0, gb_switch_delay = 0, gb_cart_switch_enabled = 0}, {control_id = 2,
                cont = 0x712b9c50 <g_dev+52178000>, tpk = 0x712b9d08 <g_dev+52178184>, last_input = 0, last_pak_type = 1, main_switch_pak = 0x0, pak_switch_delay = 0,
                gb_switch_delay = 0, gb_cart_switch_enabled = 0}, {control_id = 3, cont = 0x712b9c68 <g_dev+52178024>, tpk = 0x712b9d1c <g_dev+52178204>, last_input = 0,
                last_pak_type = 1, main_switch_pak = 0x0, pak_switch_delay = 0, gb_switch_delay = 0, gb_cart_switch_enabled = 0}}
            mpk_storages = {{
                data = 0x712bbbb0 <saved_memory+2048> "\201\001\002\003\004\005\006\a\b\t\n\v\f\r\016\017\020\021\022\023\024\025\026\027\030\031\032\033\034\035\036\037\377\377\377\377\005\032_\023", size = 32768, filename = 0x6488b354 "\260\273+q"}, {
                data = 0x712c3bb0 <saved_memory+34816> "\201\001\002\003\004\005\006\a\b\t\n\v\f\r\016\017\020\021\022\023\024\025\026\027\030\031\032\033\034\035\036\037\377\377\377\377\005\032_\023", size = 32768, filename = 0x6488b354 "\260\273+q"}, {
                data = 0x712cbbb0 <saved_memory+67584> "\201\001\002\003\004\005\006\a\b\t\n\v\f\r\016\017\020\021\022\023\024\025\026\027\030\031\032\033\034\035\036\037\377\377\377\377\005\032_\023", size = 32768, filename = 0x6488b354 "\260\273+q"}, {
                data = 0x712d3bb0 <saved_memory+100352> "\201\001\002\003\004\005\006\a\b\t\n\v\f\r\016\017\020\021\022\023\024\025\026\027\030\031\032\033\034\035\036\037\377\377\377\377\005\032_\023", size = 32768, filename = 0x6488b354 "\260\273+q"}}
            mpk = {
              data = 0x712bbbb0 <saved_memory+2048> "\201\001\002\003\004\005\006\a\b\t\n\v\f\r\016\017\020\021\022\023\024\025\026\027\030\031\032\033\034\035\036\037\377\377\377\377\005\032_\023", size = 131072, filename = 0x0}
            gbcam_backend = 0x0
            igbcam_backend = 0x0
            flashram_type = 12714014
            dd_rtc_iclock = 0x0
            dd_idisk = 0x0
            joybus_devices = {0x712b9c20 <g_dev+52177952>, 0x712b9c38 <g_dev+52177976>, 0x712b9c50 <g_dev+52178000>, 0x712b9c68 <g_dev+52178024>, 0x712b9f90 <g_dev+52178832>}
            ijoybus_devices = {0x64aa9d1c <g_ijoybus_device_controller>, 0x64aa9d1c <g_ijoybus_device_controller>, 0x64aa9d1c <g_ijoybus_device_controller>,
              0x64aa9d1c <g_ijoybus_device_controller>, 0x64aa9cf8 <g_ijoybus_device_cart>}
    #10 0x6497a0b4 in CoreDoCommand (Command=M64CMD_EXECUTE, ParamInt=0, ParamPtr=0x0) at mupen64plus-core/src/api/frontend.c:176
            rval = M64ERR_SUCCESS
            keysym = 0
            keymod = 0
    #11 0x64a53098 in EmuThreadFunction () at libretro/libretro.c:322
    No locals.
    #12 0x64a7518c in co_switch (handle=<error reading variable: Cannot access memory at address 0xfffffff0>) at libretro-common/libco/armeabi.c:101
            co_previous_handle = <error reading variable co_previous_handle (Cannot access memory at address 0xfffffff8)>

Original Issue: https://github.com/libretro/mupen64plus-libretro-nx/issues/31#issue-415177491

### PR DESCRIPTION
**Pull request by flagrama**
Fixes trying to assign to a null pointer on hard reset
Should fix #639

### CODE REVIEW

**Comment by hhromic** [comment]
Hi, thanks for your PR. I'm one of the early reporters of the segfault-on-reset in the libretro version.
I tested your fix and unfortunately doesn't work for me. It still crashes when doing a hard-reset.
Therefore it doesn't fix #639 

For what is worth, the libretro version solved the issue by doing a soft-reset instead:
https://github.com/libretro/mupen64plus-libretro-nx/commit/76612f10d5b1f94514ae65219466b5d6258aafd8

This solution have been working fine so far.

**Comment by flagrama** [comment]
Interesting. Both dynamic recompiler and pure interpreter hard reset fine for me with this change. I'm running it on Debian Linux but I'm not sure that should be an issue here. What OS did you try it on?

**Comment by flagrama** [comment]
It works fine on mupen64plus-libretro-nx master, crashes with the following on develop; 
```
Starting program: /usr/bin/retroarch -L mupen64plus_next_libretro.so ~/Public/Emu/ROM/Legend\ of\ Zelda\,\ The\ -\ Ocarina\ of\ Time\ \(USA\).n64
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7fffec9bc700 (LWP 26320)]
[New Thread 0x7fffdbf9b700 (LWP 26321)]
[New Thread 0x7fffd9521700 (LWP 26322)]
[New Thread 0x7fffd8d20700 (LWP 26323)]
[New Thread 0x7fffd851f700 (LWP 26324)]
[New Thread 0x7fffb751f700 (LWP 26325)]
[New Thread 0x7fffb6d1e700 (LWP 26326)]
[New Thread 0x7fffb651d700 (LWP 26327)]
[New Thread 0x7fffb5d1c700 (LWP 26328)]
[New Thread 0x7fffb551b700 (LWP 26329)]
[New Thread 0x7fffb4d1a700 (LWP 26330)]
[New Thread 0x7fff7ffff700 (LWP 26333)]

Thread 1 "retroarch" received signal SIGSEGV, Segmentation fault.
0x00007fffdee1a7c0 in cached_interp_MTC0 () at mupen64plus-core/src/device/r4300/mips_instructions.def:1188
1188	        cp0_regs[CP0_INDEX_REG] = rrt32 & UINT32_C(0x8000003F);
(gdb) bt full
#0  0x00007fffdee1a7c0 in cached_interp_MTC0 () at mupen64plus-core/src/device/r4300/mips_instructions.def:1188
        r4300 = 0x7fffe86ec000 <g_dev>
        cp0_regs = 0x7fffeafed258 <g_dev+42996312>
        cp0_next_interrupt = 0x7fffeafed100 <g_dev+42995968>
        __PRETTY_FUNCTION__ = "cached_interp_MTC0"
#1  0x00007fffdeea9dec in MTC0_new (copr=13, count=20622, diff=0, pcaddr=-1543503688) at mupen64plus-core/src/device/r4300/new_dynarec/new_dynarec.c:11187
        r4300 = 0x7fffe86ec000 <g_dev>
#2  0x00007fffe8fed7a5 in g_dev () at /home/vcunningham/Public/src/mupen64plus-libretro-nx/mupen64plus_next_libretro.so
#3  0xa40000b8ffffffff in  ()
#4  0x00007fffe86ec000 in g_dev () at /home/vcunningham/Public/src/mupen64plus-libretro-nx/mupen64plus_next_libretro.so
#5  0x0000555556cfc340 in  ()
#6  0x0000080000000800 in  ()
#7  0x00007fffddd23e20 in  ()
#8  0x00007fffdee79240 in dyna_linkerPython Exception <class 'gdb.MemoryError'> Cannot access memory at address 0xffffffffffffffc8: 
(gdb) info threads
 (src=  Id   Target Id                                          Frame 
* 1    Thread 0x7fffec9f2ec0 (LWP 26319) "retroarch"      0x00007fffdee1a7c0 in cached_interp_MTC0 () at mupen64plus-core/src/device/r4300/mips_instructions.def:1188
  2    Thread 0x7fffec9bc700 (LWP 26320) "retroarch"      futex_wait_cancelable (private=0, expected=0, futex_word=0x555555a678bc) at ../sysdeps/unix/sysv/linux/futex-internal.h:88
  3    Thread 0x7fffdbf9b700 (LWP 26321) "retroarch:cs0"  futex_wait_cancelable (private=0, expected=0, futex_word=0x555555f9affc) at ../sysdeps/unix/sysv/linux/futex-internal.h:88
  4    Thread 0x7fffd9521700 (LWP 26322) "retroar:disk$0" futex_wait_cancelable (private=0, expected=0, futex_word=0x555555f9f158) at ../sysdeps/unix/sysv/linux/futex-internal.h:88
  5    Thread 0x7fffd8d20700 (LWP 26323) "retroarch:sh0"  futex_wait_cancelable (private=0, expected=0, futex_word=0x555555f9c3e4) at ../sysdeps/unix/sysv/linux/futex-internal.h:88
  6    Thread 0x7fffd851f700 (LWP 26324) "retroarch:sh1"  futex_wait_cancelable (private=0, expected=0, futex_word=0x555555f9c3e0) at ../sysdeps/unix/sysv/linux/futex-internal.h:88
  7    Thread 0x7fffb751f700 (LWP 26325) "retroarch:sh2"  futex_wait_cancelable (private=0, expected=0, futex_word=0x555555f9c3e4) at ../sysdeps/unix/sysv/linux/futex-internal.h:88
  8    Thread 0x7fffb6d1e700 (LWP 26326) "retroarch:sh3"  futex_wait_cancelable (private=0, expected=0, futex_word=0x555555f9c3e0) at ../sysdeps/unix/sysv/linux/futex-internal.h:88
  9    Thread 0x7fffb651d700 (LWP 26327) "retroarc:shlo0" futex_wait_cancelable (private=0, expected=0, futex_word=0x555555f9c960) at ../sysdeps/unix/sysv/linux/futex-internal.h:88
  10   Thread 0x7fffb5d1c700 (LWP 26328) "retroarc:shlo1" futex_wait_cancelable (private=0, expected=0, futex_word=0x555555f9c960) at ../sysdeps/unix/sysv/linux/futex-internal.h:88
  11   Thread 0x7fffb551b700 (LWP 26329) "retroarc:shlo2" futex_wait_cancelable (private=0, expected=0, futex_word=0x555555f9c960) at ../sysdeps/unix/sysv/linux/futex-internal.h:88
  12   Thread 0x7fffb4d1a700 (LWP 26330) "retroarc:gdrv0" futex_wait_cancelable (private=0, expected=0, futex_word=0x5555562ab2ec) at ../sysdeps/unix/sysv/linux/futex-internal.h:88
  13   Thread 0x7fff7ffff700 (LWP 26333) "threaded-ml"    0x00007ffff4391edf in __GI___poll (fds=0x7fff78007180, nfds=3, timeout=-1) at ../sysdeps/unix/sysv/linux/poll.c:29
(gdb)```

**Comment by hhromic** [comment]
Thanks for testing!
I'm using Raspbian Stretch on a Raspberry Pi 3B+.
Yes I should have mention more details of my testing setup. Indeed I'm using `mupen64plus-next` from the `develop` branch (the active dev branch) and not `master`.

Indeed interesting that it crashes with the develop branch and not master. Let me know if you need any further test or debugging output from my side to help.

**Comment by flagrama** [comment]
It seems that this fixes hard reset on the pure interpreter. It is very likely I had set this when I tried to reset on master.

The master and develop branches seem to crash the same way when using the dynamic recompiler. I'm still not running into this particular issue when building the normal mupen64plus-core built from this repo. It seems to restart properly even using dynarec, though other issues do still crop up for me after a hard reset with dynarec enabled. These might just be an issue with the GUI I'm using though.

**Comment by flagrama** [comment]
Seems like there was a lot more than just a null pointer going on. Going by how the initial run code happens, it seems to use this run_device function that hard_reset wasn't doing. Adding this call to the hard_reset handler seems to have fixed hard reset in all three emulation modes for me now.

**Comment by hhromic** [comment]
Ah nice! I'll try it asap on the mupen-next libretro core (develop branch) and report back! Thanks!

**Comment by flagrama** [comment]
The previous code seemed to be creating another instance of the emulator each time hard reset was used, so RAM usage would just continue to grow after each hard reset. These small changes seem to actually solve most of the problems.

The only thing I seem to actually be running into an issue with is what I assume would be "old dynarec" where after a hard reset closing the ROM just results in an endless loop of `Core Warning: Instruction pointer is 0 at dyna_stop()`. The three other emumodes work fine though it seems.

**Comment by hhromic** [comment]
Thanks for your work on this!

I had the chance to test now. Indeed the changes you did couple of days ago didn't work very well (games segfault after reset) but the recent changes you did yesterday seem to be working fine now with the libretro version!

I tested the dynarec, pure interpreter and cached interpreter emulation modes and the hard reset function is working as expected in all of them for me. So far I tested with couple of games and I don't notice issues.
I also tested closing the ROM and found no issues (the mupen next libretro version by default uses the new dynarec).

So in summary, thanks again for fixing this long-standing bug.

**Comment by flagrama** [comment]
Good to know this seems to be at least "good enough."

Through some more experimentation it seems that the following code will also get old dynarec to play along. Unfortunately I have no idea what these variables actually are and if persisting them over a hard reset defeats the purpose of a hard reset or not. As such I won't be adding them to this PR currently.

```C
    long long save_rsp = r4300->recomp.save_rsp;
    long long save_rip = r4300->recomp.save_rip;

    poweron_device(dev);
/* snip */
    if(r4300->emumode >= 2)
    {
#ifdef NEW_DYNAREC
        new_dynarec_cleanup();
        new_dynarec_init();
#else
        r4300->recomp.save_rsp = save_rsp;
        r4300->recomp.save_rip = save_rip;
#endif
    }
```

**Comment by bsmiles32** [comment]
Merged via PR #746 which includes the save_rsp, save_rip bits.
Thanks to all involved !

**Comment by charlemagnelasse** [comment]
This is completely wrong for non-x86-64. Here is the definition from src/device/r4300/r4300_core.h

    127 #if defined(__x86_64__)
    128         long long save_rsp;
    129         long long save_rip;
    130 
    131         /* that's where the dynarec will restart when going back from a C function */
    132         unsigned long long* return_address;
    133 #else
    134         long save_ebp;
    135         long save_ebx;
    136         long save_esi;
    137         long save_edi;
    138         long save_esp;
    139         long save_eip;
    140 
    141         /* that's where the dynarec will restart when going back from a C function */
    142         unsigned long* return_address;
    143 #endif
